### DISCUSSION TITLE
Validação de datas no C#

### DISCUSSION THREAD

**Discussion by RafaelFCouto** [balta-io/forum#452] [Tira Dúvidas] [answered]
Bom, estou trabalhando no desenvolvimento de um sistema web, utilizando o Razor Pages, preciso da seguinte validação: 

- Data de Início de um evento: Maior que a data atual;
- Data final de um evento sempre superior a inicial.

Consegui validar dessa forma para data inicial:

![image](https://github.com/balta-io/forum/assets/149386775/fa4ab252-e416-4337-8e53-cdea3b7e685e)
Usando esse construtor da Data Annotation Ranger:
![image](https://github.com/balta-io/forum/assets/149386775/c5879de5-e5a6-4032-bb7f-1006d3a21846)

Mas, esse construtor não aceita as datas como sendo variáveis, mesmo se for do tipo string... 

Alguém tem alguma luz de como fazer?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @RafaelFCouto, como vai? Tentou com as datas no formato internacional? yyyy-MM-dd

**Reply by RafaelFCouto** [reply]
Não tentei, vai ser o próximo ponto... já na questão de realizar o limite mínimo utilizando o DateTime.Now, tem alguma ideia?

**Comment by malighieri** [comment]
Para uma gama maior de validações eu costumo usar a biblioteca [MVC Foolproof Validation](https://github.com/rpgkaiser/FoolProof.Core)
Ela estende as validações por Data Annotations incluindo várias outras possibilidades.
Para o seu caso, acredito que pode usar a Data Annotation _**GreaterThan**_. Você passa para ela o nome da propriedade a qual a propriedade sendo validada precisa ter valor maior. Por exemplo:

```
public class EventoViewModel
{
    [Required]
    public string Nome { get; set; }

    public DateTime Agora => DateTime.Now;

    [Required]
    [GreaterThan("Agora")]
    public DateTime Inicio { get; set; }

    [Required]
    [GreaterThan("Inicio")]
    public DateTime Fim { get; set; }
}
```

O exemplo acima foi inspirado no exemplo da própria [Documentação do MVC Foolproof Validation](https://codeplexarchive.org/project/foolproof)
Alterei os nomes das propriedades para português além de modificar a classe para contemplar uma solução para validar se uma data é maior que a data atual. Assim temos:

1. [GreaterThan("Agora")]
   Indica que a propriedade sendo validada tem que ter valor maior que a propriedade _**Agora**_. No exemplo criei uma propriedade _**Agora**_ cujo valor é dado por DateTime.Now. Sendo assim, a propriedade sendo validada sempre deverá ter valor maior do que DateTime.Now.

2. [GreaterThan("Inicio")]
   Indica que a propriedade sendo validada tem que ter valor maior que a propriedade _**Inicio**_. No exemplo a propriedade _**Inicio**_ tem o valor informado pelo usuário, e a propriedade sendo validada terá que ter o valor maior que ela.

O construtor desta Data Annotation permite configurar a mensagem que será retornada bem como deixar passar a validação caso valor da propriedade seja null
![image](https://github.com/balta-io/forum/assets/23501682/1774d7cb-5af2-4d8e-95fc-0f8f8e714801)

Além das Data Annotations contidas na biblioteca, a documentação provê exemplos de como você implementar sua própria validação se necessário.

Por fim, esta biblioteca tem também uma biblioteca acessória em javascript para que você possa levar várias destas validações para o browser, sem precisar fazer o post para que elas sejam feitas. Ela ainda checará quando fizer o post. Você consegue encontrar como utilizar ela na documentação.

**Reply by RafaelFCouto** [reply]
Muito obrigado pelo suporte, a nível de código não apresentou nenhum erro...
Mas, quando fui pra execução e entrei na page que estava utilizando esse estilo de código deu esse erro: 

![image](https://github.com/balta-io/forum/assets/149386775/b59510cd-575f-45ea-9f89-3f9880475803)

Tentei realizar pesquisas para encontrar uma solução, mas infelizmente não obtive sucesso. 

Tu tens mais alguma luz? Preciso de alguma configuração na class program.cs? ou algo do tipo? 

Grato desde já pela compreensão, paciência e suporte.
